File: golem-common/src/base_model/oplog/multipart.rs
// Copyright 2024-2026 Golem Cloud
//
// Licensed under the Golem Source License v1.1 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://license.golem.cloud/LICENSE
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

/// A parsed part from a multipart/mixed message.
#[derive(Debug)]
pub struct MultipartPart<'a> {
    /// The `name` parameter from Content-Disposition, if present.
    pub name: Option<String>,
    /// The Content-Type header value, if present.
    pub content_type: Option<String>,
    /// The raw body bytes of this part.
    pub body: &'a [u8],
}

/// Extracts the boundary parameter from a `multipart/mixed; boundary=...` content-type string.
pub fn extract_boundary(content_type: &str) -> Option<&str> {
    let ct = content_type.strip_prefix("multipart/mixed")?;
    let rest = ct.trim_start();
    let rest = rest.strip_prefix(';')?;
    let rest = rest.trim_start();
    for param in rest.split(';') {
        let param = param.trim();
        if let Some(value) = param.strip_prefix("boundary=") {
            let value = value.trim();
            // Remove optional quotes
            if value.starts_with('"') && value.ends_with('"') && value.len() >= 2 {
                return Some(&value[1..value.len() - 1]);
            }
            return Some(value);
        }
    }
    None
}

/// Parses a multipart/mixed body into its constituent parts.
///
/// The `boundary` should be extracted from the Content-Type header using `extract_boundary`.
/// Returns `None` if the body cannot be parsed.
pub fn parse_multipart_mixed<'a>(boundary: &str, data: &'a [u8]) -> Option<Vec<MultipartPart<'a>>> {
    let delimiter = format!("--{boundary}");
    let close_delimiter = format!("--{boundary}--");

    let delimiter_bytes = delimiter.as_bytes();
    let close_delimiter_bytes = close_delimiter.as_bytes();

    let mut parts = Vec::new();
    let mut pos = 0;

    // Skip preamble: find the first delimiter
    pos = find_bytes(data, pos, delimiter_bytes)?;
    pos += delimiter_bytes.len();

    // Skip CRLF or LF after delimiter
    pos = skip_line_ending(data, pos);

    loop {
        // Check if we've hit the close delimiter
        if data[pos..].starts_with(close_delimiter_bytes) {
            break;
        }

        // Find the end of headers (empty line)
        let header_end = find_empty_line(data, pos)?;
        let headers_slice = &data[pos..header_end];
        let headers_str = std::str::from_utf8(headers_slice).ok()?;

        // Parse headers
        let mut name = None;
        let mut content_type = None;
        for line in headers_str.lines() {
            let line = line.trim();
            if let Some(ct) = line.strip_prefix("Content-Type:") {
                content_type = Some(ct.trim().to_string());
            } else if let Some(cd) = line.strip_prefix("Content-Disposition:") {
                // Extract name="..." parameter
                if let Some(name_start) = cd.find("name=\"") {
                    let rest = &cd[name_start + 6..];
                    if let Some(name_end) = rest.find('"') {
                        name = Some(rest[..name_end].to_string());
                    }
                }
            }
        }

        // Body starts after the empty line
        let body_start = skip_empty_line(data, header_end);

        // Find the next delimiter
        let body_end = find_bytes(data, body_start, delimiter_bytes)?;

        // Strip trailing CRLF or LF before the delimiter
        let body_end = strip_trailing_line_ending(data, body_start, body_end);

        parts.push(MultipartPart {
            name,
            content_type,
            body: &data[body_start..body_end],
        });

        // Move past the delimiter
        pos = body_end;
        // Skip the line ending before the delimiter
        pos = skip_line_ending(data, pos);
        pos = find_bytes(data, pos, delimiter_bytes)?;
        pos += delimiter_bytes.len();
        // Check for close delimiter suffix
        if data.get(pos..pos + 2) == Some(b"--") {
            break;
        }
        pos = skip_line_ending(data, pos);
    }

    Some(parts)
}

fn find_bytes(haystack: &[u8], start: usize, needle: &[u8]) -> Option<usize> {
    haystack[start..]
        .windows(needle.len())
        .position(|w| w == needle)
        .map(|p| p + start)
}

fn skip_line_ending(data: &[u8], pos: usize) -> usize {
    if data.get(pos..pos + 2) == Some(b"\r\n") {
        pos + 2
    } else if data.get(pos) == Some(&b'\n') {
        pos + 1
    } else {
        pos
    }
}

fn find_empty_line(data: &[u8], start: usize) -> Option<usize> {
    // Look for \r\n\r\n or \n\n
    for i in start..data.len().saturating_sub(1) {
        if data[i] == b'\n' && data[i + 1] == b'\n' {
            return Some(i);
        }
        if i + 3 < data.len()
            && data[i] == b'\r'
            && data[i + 1] == b'\n'
            && data[i + 2] == b'\r'
            && data[i + 3] == b'\n'
        {
            return Some(i);
        }
    }
    None
}

fn skip_empty_line(data: &[u8], pos: usize) -> usize {
    if data.get(pos..pos + 4) == Some(b"\r\n\r\n") {
        pos + 4
    } else if data.get(pos..pos + 2) == Some(b"\n\n") {
        pos + 2
    } else {
        pos
    }
}

fn strip_trailing_line_ending(data: &[u8], body_start: usize, body_end: usize) -> usize {
    if body_end >= body_start + 2 && data[body_end - 2] == b'\r' && data[body_end - 1] == b'\n' {
        body_end - 2
    } else if body_end > body_start && data[body_end - 1] == b'\n' {
        body_end - 1
    } else {
        body_end
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use test_r::test;

    #[test]
    fn test_extract_boundary() {
        assert_eq!(
            extract_boundary("multipart/mixed; boundary=abc123"),
            Some("abc123")
        );
        assert_eq!(
            extract_boundary("multipart/mixed; boundary=\"abc-123\""),
            Some("abc-123")
        );
        assert_eq!(extract_boundary("application/json"), None);
    }

    #[test]
    fn test_parse_simple_multipart() {
        let boundary = "test-boundary";
        let body = format!(
            "--{boundary}\r\n\
             Content-Type: application/json\r\n\
             Content-Disposition: attachment; name=\"state\"\r\n\
             \r\n\
             {{\"key\":\"value\"}}\r\n\
             --{boundary}\r\n\
             Content-Type: application/x-sqlite3\r\n\
             Content-Disposition: attachment; name=\"db:main\"\r\n\
             \r\n\
             SQLITEDATA\r\n\
             --{boundary}--\r\n"
        );

        let parts = parse_multipart_mixed(boundary, body.as_bytes()).unwrap();
        assert_eq!(parts.len(), 2);

        assert_eq!(parts[0].name.as_deref(), Some("state"));
        assert_eq!(parts[0].content_type.as_deref(), Some("application/json"));
        assert_eq!(parts[0].body, b"{\"key\":\"value\"}");

        assert_eq!(parts[1].name.as_deref(), Some("db:main"));
        assert_eq!(
            parts[1].content_type.as_deref(),
            Some("application/x-sqlite3")
        );
        assert_eq!(parts[1].body, b"SQLITEDATA");
    }

    #[test]
    fn test_parse_multipart_with_binary() {
        let boundary = "uuid-boundary-12345";
        let json_part = b"{\"version\":1}";
        let binary_part: &[u8] = &[0x00, 0xFF, 0x53, 0x51, 0x4C];

        let mut body = Vec::new();
        body.extend_from_slice(format!("--{boundary}\r\n").as_bytes());
        body.extend_from_slice(b"Content-Type: application/json\r\n");
        body.extend_from_slice(b"Content-Disposition: attachment; name=\"state\"\r\n");
        body.extend_from_slice(b"\r\n");
        body.extend_from_slice(json_part);
        body.extend_from_slice(format!("\r\n--{boundary}\r\n").as_bytes());
        body.extend_from_slice(b"Content-Type: application/x-sqlite3\r\n");
        body.extend_from_slice(b"Content-Disposition: attachment; name=\"db:cache\"\r\n");
        body.extend_from_slice(b"\r\n");
        body.extend_from_slice(binary_part);
        body.extend_from_slice(format!("\r\n--{boundary}--\r\n").as_bytes());

        let parts = parse_multipart_mixed(boundary, &body).unwrap();
        assert_eq!(parts.len(), 2);
        assert_eq!(parts[0].name.as_deref(), Some("state"));
        assert_eq!(parts[0].body, json_part);
        assert_eq!(parts[1].name.as_deref(), Some("db:cache"));
        assert_eq!(parts[1].body, binary_part);
    }
}

File: sdks/ts/packages/golem-ts-sdk/src/internal/multipart.ts
// Copyright 2024-2026 Golem Cloud
//
// Licensed under the Golem Source License v1.1 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://license.golem.cloud/LICENSE
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

const CRLF = '\r\n';

const textEncoder = new TextEncoder();
const textDecoder = new TextDecoder();

export interface MultipartPart {
  name: string;
  contentType: string;
  body: Uint8Array;
}

function generateBoundary(): string {
  if (typeof crypto !== 'undefined' && typeof crypto.randomUUID === 'function') {
    return crypto.randomUUID().replace(/-/g, '');
  }
  let hex = '';
  for (let i = 0; i < 32; i++) {
    hex += Math.floor(Math.random() * 16).toString(16);
  }
  return hex;
}

function containsBoundary(parts: MultipartPart[], boundary: string): boolean {
  const marker = textEncoder.encode(`${CRLF}--${boundary}`);
  for (const part of parts) {
    if (indexOf(part.body, marker) !== -1) {
      return true;
    }
  }
  return false;
}

function indexOf(haystack: Uint8Array, needle: Uint8Array): number {
  if (needle.length === 0) return 0;
  if (needle.length > haystack.length) return -1;
  outer: for (let i = 0; i <= haystack.length - needle.length; i++) {
    for (let j = 0; j < needle.length; j++) {
      if (haystack[i + j] !== needle[j]) continue outer;
    }
    return i;
  }
  return -1;
}

function concat(chunks: Uint8Array[]): Uint8Array {
  let total = 0;
  for (const c of chunks) total += c.length;
  const result = new Uint8Array(total);
  let offset = 0;
  for (const c of chunks) {
    result.set(c, offset);
    offset += c.length;
  }
  return result;
}

export function encodeMultipart(parts: MultipartPart[]): {
  data: Uint8Array;
  boundary: string;
} {
  let boundary = generateBoundary();
  while (containsBoundary(parts, boundary)) {
    boundary = generateBoundary();
  }

  const chunks: Uint8Array[] = [];

  for (const part of parts) {
    const header =
      `--${boundary}${CRLF}` +
      `Content-Type: ${part.contentType}${CRLF}` +
      `Content-Disposition: attachment; name="${part.name}"${CRLF}` +
      CRLF;
    chunks.push(textEncoder.encode(header));
    chunks.push(part.body);
    chunks.push(textEncoder.encode(CRLF));
  }

  chunks.push(textEncoder.encode(`--${boundary}--${CRLF}`));

  return { data: concat(chunks), boundary };
}

export function decodeMultipart(data: Uint8Array, boundary: string): MultipartPart[] {
  const delimiter = textEncoder.encode(`${CRLF}--${boundary}`);
  const crlfBytes = textEncoder.encode(CRLF);
  const lfByte = 0x0a; // \n

  // Normalise: strip leading CRLF or LF if present
  let start = 0;
  if (data.length >= 2 && data[0] === 0x0d && data[1] === 0x0a) {
    start = 2;
  } else if (data.length >= 1 && data[0] === 0x0a) {
    start = 1;
  }

  // The first boundary appears without a leading CRLF
  const firstDelimiter = textEncoder.encode(`--${boundary}`);

  // Verify the body starts with the opening boundary
  if (!startsWith(data, firstDelimiter, start)) {
    throw new Error('Multipart body does not start with boundary');
  }

  // Split the body on the delimiter
  const rawSections = splitOnDelimiter(data, delimiter, start + firstDelimiter.length);

  const parts: MultipartPart[] = [];
  const seenNames = new Set<string>();

  for (const section of rawSections) {
    // Skip the closing delimiter section
    if (startsWith(section, textEncoder.encode('--'), 0)) {
      continue;
    }

    // Section starts after delimiter; skip the CRLF or LF that follows it
    let sectionStart = 0;
    if (section.length >= 2 && section[0] === 0x0d && section[1] === 0x0a) {
      sectionStart = 2;
    } else if (section.length >= 1 && section[0] === 0x0a) {
      sectionStart = 1;
    }

    // Find the blank line separating headers from body (CRLF CRLF or LF LF)
    const headerEndCrlf = findDoubleNewline(section, sectionStart, crlfBytes);
    const headerEndLf = findDoubleNewlineLf(section, sectionStart);

    let headerEnd: number;
    let bodyStart: number;

    if (headerEndCrlf !== -1 && (headerEndLf === -1 || headerEndCrlf <= headerEndLf)) {
      headerEnd = headerEndCrlf;
      bodyStart = headerEndCrlf + 4; // skip \r\n\r\n
    } else if (headerEndLf !== -1) {
      headerEnd = headerEndLf;
      bodyStart = headerEndLf + 2; // skip \n\n
    } else {
      throw new Error('Could not find end of headers in multipart part');
    }

    const headerText = textDecoder.decode(section.slice(sectionStart, headerEnd));
    const headerLines = headerText.split(/\r?\n/);

    let contentType = '';
    let name = '';

    for (const line of headerLines) {
      const colonIdx = line.indexOf(':');
      if (colonIdx === -1) continue;
      const key = line.slice(0, colonIdx).trim().toLowerCase();
      const value = line.slice(colonIdx + 1).trim();

      if (key === 'content-type') {
        contentType = value;
      } else if (key === 'content-disposition') {
        const nameMatch = value.match(/name="([^"]+)"/);
        if (nameMatch) {
          name = nameMatch[1];
        }
      }
    }

    if (!name) {
      throw new Error('Multipart part missing name in Content-Disposition');
    }
    if (seenNames.has(name)) {
      throw new Error(`Duplicate multipart part name: ${name}`);
    }
    seenNames.add(name);

    // Body: everything after headers until end of section.
    // Strip trailing CRLF or LF that precedes the next delimiter.
    let bodyEnd = section.length;
    if (bodyEnd >= 2 && section[bodyEnd - 2] === 0x0d && section[bodyEnd - 1] === 0x0a) {
      bodyEnd -= 2;
    } else if (bodyEnd >= 1 && section[bodyEnd - 1] === lfByte) {
      bodyEnd -= 1;
    }

    const body = section.slice(bodyStart, bodyEnd);

    parts.push({ name, contentType, body });
  }

  return parts;
}

function startsWith(data: Uint8Array, prefix: Uint8Array, offset: number): boolean {
  if (offset + prefix.length > data.length) return false;
  for (let i = 0; i < prefix.length; i++) {
    if (data[offset + i] !== prefix[i]) return false;
  }
  return true;
}

function splitOnDelimiter(data: Uint8Array, delimiter: Uint8Array, start: number): Uint8Array[] {
  const sections: Uint8Array[] = [];
  let pos = start;

  while (pos < data.length) {
    const next = indexOf(data.slice(pos), delimiter);
    if (next === -1) {
      sections.push(data.slice(pos));
      break;
    }
    sections.push(data.slice(pos, pos + next));
    pos = pos + next + delimiter.length;
  }

  return sections;
}

function findDoubleNewline(data: Uint8Array, start: number, crlf: Uint8Array): number {
  // Look for \r\n\r\n
  for (let i = start; i <= data.length - 4; i++) {
    if (
      data[i] === crlf[0] &&
      data[i + 1] === crlf[1] &&
      data[i + 2] === crlf[0] &&
      data[i + 3] === crlf[1]
    ) {
      return i;
    }
  }
  return -1;
}

function findDoubleNewlineLf(data: Uint8Array, start: number): number {
  // Look for \n\n
  for (let i = start; i <= data.length - 2; i++) {
    if (data[i] === 0x0a && data[i + 1] === 0x0a) {
      return i;
    }
  }
  return -1;
}
